Virtual keyboard

kkulahin opened the pull request:
1. Task - [RSS Virtual keyboard](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/codejam-virtual-keyboard.md)
2. Screenshot: ![virtual-keyboard](https://user-images.githubusercontent.com/60391325/79041513-2bcab700-7bf9-11ea-988c-f3db3e00f755.PNG)
3. Demo - [RSS Virtual keyboard demo](https://blacktruth.github.io/virtual-keyboard/)
4. Demo for mentor ( commits after deadline) [RSS Virtual keyboard demo 2.0](https://blacktruth-virtual-keyboard.netlify.com/dist/index.html)
5. Date of completion - 07.04.2020 / date of deadline 07.04.2020
6. Total

Task ( 8 / 8 )
Minimum set. ( 2 / 2 )
- [+] Implemented the generation of DOM element in index.html <body /> initially empty.
- [+] Pressing the button on the physical keyboard highlights the button on the virtual one (check the pressing of numbers, letters, punctuation, backspace, del (if present), enter, shift, alt, ctrl, tab, caps lock, space, down-up arrows - left-right)

Standart set. ( 2 / 2 )
- [+] There is a switch between the Russian and English languages ​​(the key combination for switching the language should be indicated on the page with the keyboard), as well as saving the selected language.
- [+] Mouse clicks on the buttons on the virtual keyboard or clicking on the buttons on the physical keyboard inserts the characters to input (textarea).

Additional set. ( 1 / 1 )
- [+] Implemented button click animation.

Technical requirements. ( 3 / 3 )
- [+] Usage of ES6 and higher chips in the code: arrow functions, destructurisation
- [+] Eslint usage
- [+] The requirements for the repository, commits and PR are met.

Penalty. ( 0 / 1 )
- [-] Errors in the console related to the executable code (errors like favicon.ico: Failed to load resource: the server responded with a status of 404 are not taken into account) or eslint-config-airbnb-base warnings

Mark calculation: 110

kkulahin committed:
feat: add textarea and first line of keyboard

kkulahin committed:
feat: add second keyboard line

kkulahin committed:
fix: add focus on timeout

kkulahin committed:
feat: add keydown and keyup events

kkulahin committed:
feat: add change layout on buttons press

kkulahin committed:
feat: add arrow up and down functions

kkulahin committed:
feat: add en/ru languages for the rest of keyboard, add animation

kkulahin committed:
fix: alt, ctrl event, focus on textarea, caps lock behavior, tab default behavior

kkulahin committed:
feat: add eslint and resolved issues

kkulahin committed:
refactor: move script position

kkulahin committed:
fix: destructurization add

kkulahin committed:
fix: prevent alt, fix caps behavior

kkulahin committed:
fix: scripts path

kkulahin committed:
fix: add switch layout message

Andrew-Bogolubof commented on .eslintrc.js:
```diff
@@ -0,0 +1,19 @@
+        "quotes": ["error", "double"],
+        "linebreak-style" : ["error", "windows"]
+    }
+};
```
new line

kkulahin replied:
Add new line at the end of the file in the commit below:
https://github.com/BlackTruth/virtual-keyboard/commit/bace22c72e918a327db89ea7670c0dcf87c2bd6f?diff=unified#diff-e4403a877d80de653400d88d85e4801a

Andrew-Bogolubof commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+
```
why empty string?

kkulahin replied:
https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/.gitignore#L1
Fixed

Andrew-Bogolubof commented on assets/styles.css:
```diff
@@ -0,0 +1,129 @@
+  margin-top: 20px;
+}
+
+.keyboard-panel {
```
empty style

kkulahin replied:
Removed empty styles in commit below

https://github.com/BlackTruth/virtual-keyboard/commit/88ba0949a558aa2d031f3d98f72f4d8880f72717

Andrew-Bogolubof commented on assets/styles.css:
```diff
@@ -0,0 +1,129 @@
+  cursor: pointer;
+  background: #DBE6F0;
+}
+/* .keyboard-line__button:active, */
```
No comments in the completed task.

kkulahin replied:
Remove commented style in commit below

https://github.com/BlackTruth/virtual-keyboard/commit/155ce0b54e3bcf888457e882c53fcf5462807995#diff-42cb364f9d4783c39009483c0ad61341

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+  caps: false,
+};
+
+if (!window.localStorage.getItem("virtualKeyBoardLang")) window.localStorage.setItem("virtualKeyBoardLang", "en");
```
It's really hard to read such conditions. Better
`if(your condition) {    

    code here

}`

kkulahin replied:
Moved these lines to class and refactored
https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/classes/VirtualKeyboard.js#L21-L24

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+
+if (!window.localStorage.getItem("virtualKeyBoardLang")) window.localStorage.setItem("virtualKeyBoardLang", "en");
+
+const vKeyBoard = {
```
Isn't it better to create a class, that accept textarea as input, and has methods to focus and insert symbols?

kkulahin replied:
Created class which facilitates usage of texarea

https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/classes/VirtualKeyboardInput.js#L1-L5

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+keyboardPanel.className = "keyboard-panel";
+keyboard.appendChild(keyboardPanel);
+
+// structure of key object:
```
No comments in completed task.

kkulahin replied:
Deleted commented code

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+// };
+
+const arrayOfButtonKeys = [];
+arrayOfButtonKeys.push([
```
All this configurations is better to move to another files.

kkulahin replied:
Configuration was moved to another file.

https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/keyboardConstants.js#L1-L8

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+  },
+]);
+
+function getType(key) {
```
Functions, that just help to do smth is better to move to another file like helper.js or utils.js.

kkulahin replied:
Helper function was moved to another file.

https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/helper.js#L1-L5

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+      shiftedText.innerText = getChar(key, "shifted");
+    }
+
+    keyDom.addEventListener("mousedown", () => {
```
As I see you add event listener for every key separately - it is not very good approach. Event delegation is better. You just delegate event handling to the parent element of all keys and handle events in event listener by getting from event button, that was clicked.

kkulahin replied:
Click and mouse down events assigned only to the keyboard panel instead of each key button and was moved to class.

https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/classes/VirtualKeyboard.js#L70-L73
https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/classes/VirtualKeyboard.js#L335

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+const FUNCTIONAL_KEYS = {
```
Move constants to another file.

kkulahin replied:
The state of functional keys was moved to class property.

https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/classes/VirtualKeyboard.js#L7-L20

Andrew-Bogolubof commented on script.js:
```diff
@@ -0,0 +1,841 @@
+          if (c !== "keyboard-line__button") button.classList.remove(c);
+        });
+        const mainText = button.querySelector(
+          ".keyboard-line__button-main-text",
```
For this strings it's better to create some object of constants, because you can make some mistake in ther word, and it will be really hard to find the mistake.

kkulahin replied:
All these strings was moved to styleConstants.js

https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/styleConstants.js#L1-L11

Andrew-Bogolubof commented:
Overall, code must not be written in only one file. It must be separated by functionality it provides.

Andrew-Bogolubof commented:
![image](https://user-images.githubusercontent.com/47366675/79037385-21002a00-7bd9-11ea-86e8-47eb9942fe15.png)
Have you ever seen such configuration of arrows buttons?))
This is KeyBoard on Windows: 
![image](https://user-images.githubusercontent.com/47366675/79037420-66245c00-7bd9-11ea-940a-7b63f8893ac6.png)

Andrew-Bogolubof commented:
I cannot switch language by cliscking just shift + alt on my keyboard.

Andrew-Bogolubof commented:
For now I can grade your task for 78 points.

Andrew-Bogolubof commented on index.html:
```diff
@@ -0,0 +1,15 @@
+      rel="stylesheet"
+    />
+    <link rel="stylesheet" href="assets/styles.css" />
+    <script type="text/javascript" src="script.js" async></script>
```
If you work in script with DOM better to move it to the end of the body. Read about it [learjavascript](https://learn.javascript.ru/script-async-defer)

kkulahin replied:
In the beginning i had this part in correct place, but then I thought it can be misunderstood by cross-chekers hence I moved it in head)) But now I have fixed.
https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/dist/index.html#L13-L15

Andrew-Bogolubof commented:
Let me give you piece of advice - read about MVC model)

kkulahin committed:
refactor: remove empty styles

kkulahin committed:
fix: move arrow-down under arrow-up

kkulahin committed:
refactor: moved <script /> inside <body />

kkulahin committed:
refactor: remove empty line from gitignore file

kkulahin committed:
refactor: remove empty line from the end of the .eslintrc.js

kkulahin committed:
feat: add html and css linters

kkulahin committed:
refactor: run style linter on assets/style.css

kkulahin committed:
feat: split script.js to several modules

kkulahin committed:
feat: add webpack

kkulahin committed:
fix: change layout on phisical key press

kkulahin committed:
fix: right alt on change layout

kkulahin committed:
refactor: move getType & getChar functions to helper.js

kkulahin committed:
fix: move event listener from keys to panel

add styleConstants.js to store names of css classes
add VirtualKeyboardInput.js to store class
created helper function getKeyModule

kkulahin committed:
refactor: add class VirtualKeyboard

moved component creation from script.js to VirtualKeyboard.js

kkulahin committed:
refactor: use eslint and resolve problems

kkulahin committed:
fix: windows button key press bug

kkulahin commented:
> ![image](https://user-images.githubusercontent.com/47366675/79037385-21002a00-7bd9-11ea-86e8-47eb9942fe15.png)
> Have you ever seen such configuration of arrows buttons?))
> This is KeyBoard on Windows:
> ![image](https://user-images.githubusercontent.com/47366675/79037420-66245c00-7bd9-11ea-940a-7b63f8893ac6.png)

Position of arrow keys was fixed
![Arrows](https://user-images.githubusercontent.com/60391325/79069365-55a5db80-7cd6-11ea-9dda-5cb5a6341cf4.PNG)

kkulahin commented:
> I cannot switch language by cliscking just shift + alt on my keyboard.

Change layout in real keyboard was fixed. I previously released such behavior for capslock and forgot to check language switch. And it all worked fine for me because in my system I have alt+shift as switch layout combination :)

kkulahin commented:
Published new version to netify since new commits is not allowed in gh-pages after deadline:
[Link to Virtual Keyboard 2.0](https://blacktruth-virtual-keyboard.netlify.com/dist/index.html)

kkulahin committed:
refactor: add new line in the end of .gitignore

kkulahin committed:
refactor: add new line in the end of files
